Fix #20399: Added tests for POST, PUT and DELETE and error handling for invalid method using 405 status code.

mon4our opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #20399 .
2. This PR does the following: Adds tests for POST, PUT and DELETE similar to GET (test_that_no_get_results_in_500_error). Adds error handling for invalid method using 405 status code (If you send a post request to an endpoint that only supports get, you get a 405 error). 
3. (For bug-fixing PRs only) The original bug occurred because: The required test was only written for GET. When an inavlid method was sent, a 500 status code was given instead of 405.

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).
## Proof that changes are correct
Before:
- Test for no 500 errors only for GET: 
![image](https://github.com/oppia/oppia/assets/120438312/8ff1c21e-b5cc-4088-87b8-236fca9a1104)

- Sending POST (PUT/DELETE) request to an endpoint which only supports GET results in 500 error:
![Screenshot 2024-06-21 023918](https://github.com/oppia/oppia/assets/120438312/36a700f4-2045-443e-b0a4-db0dca85deed)
```
)
]
}'
{
"error": "post() got an unexpected keyword argument 'opportunity_type'",
"status_code": 500
}
``` 
After:
- Test for no 500 errors for all requests:
![image](https://github.com/oppia/oppia/assets/120438312/8ff1c21e-b5cc-4088-87b8-236fca9a1104)
![Screenshot 2024-06-22 132254](https://github.com/oppia/oppia/assets/120438312/b2ebc290-356f-41ab-a227-3e0c3af2ba6a)
![Screenshot 2024-06-22 132300](https://github.com/oppia/oppia/assets/120438312/b6bd3616-d777-49a3-9650-13b636f09936)
![Screenshot 2024-06-22 132326](https://github.com/oppia/oppia/assets/120438312/f7d9b652-865c-456c-b159-0af5a47f2e86)

- Sending POST (PUT/DELETE) request to an endpoint which only supports GET results in 405 error instead of 500:
![Screenshot 2024-06-21 023918](https://github.com/oppia/oppia/assets/120438312/36a700f4-2045-443e-b0a4-db0dca85deed)
```
)
]
}'
{
"error": "Invalid method. ContributionOpportunitiesHandler does not have POST requests.",
"status_code": 405
}
```

- `base_test.py` runs successfully without any failures:
![image](https://github.com/oppia/oppia/assets/120438312/28da5406-0be5-4078-8c38-74e5328daefa)

- `test_utils_test.py` runs successfully without any failures:
![Screenshot 2024-06-22 013326](https://github.com/oppia/oppia/assets/120438312/e13e3f84-8728-47a9-bcc3-3d79f32c9044)



## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

mon4our committed:
Added no 500 error tests for post put and delete. Also added 405 status code handling for invalid requests.

mon4our committed:
Fixed linting issues

mon4our committed:
Fixed more linting issues

mon4our committed:
Fixed even more linting issues

mon4our committed:
Fixed MyPy issues

mon4our committed:
Fixed linting issues

mon4our commented:
@U8NWXD PTAL

mon4our committed:
Fixed backend test error and updated 405 error message.

mon4our committed:
Fixed linting errors

mon4our committed:
Fixed linting errors

vojtechjelinek commented on core/tests/test_utils.py:
```diff
@@ -3026,12 +3029,35 @@ def get_response_without_checking_for_errors(
         with self.swap(base, 'load_template', mock_load_template):
-            response = self.testapp.get(url, params=params, expect_errors=True)
+
+            if http_method == 'GET':
+                response = self.testapp.get(
+                url, params=params, expect_errors=True
+                )
```
```suggestion
            if http_method == 'GET':
                response = self.testapp.get(
                    url, params=params, expect_errors=True
                )
```

ditto below

mon4our replied:
Done

vojtechjelinek reviewed (changes requested):
Thanks! Just one formatting thing.

U8NWXD commented:
@vojtechjelinek has already reviewed for the dev workflow team, so I don't think you need my review. Feel free to re-assign me if you want my feedback though!

mon4our committed:
Added requested formatting change

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

mon4our committed:
Added requested formatting changes

mon4our committed:
Merge branch 'develop' into add-similar-tests-for-no-500-error

mon4our commented:
@vojtechjelinek PTAL

vojtechjelinek reviewed (approved):
LGTM!

lkbhitesh07 commented on core/controllers/profile_test.py:
```diff
@@ -1168,7 +1168,7 @@ def raise_exception() -> None:
                     'tag': 'Web',
                     'name': 'Name'
-                }, csrf_token=csrf_token, expected_status_int=500)
+                }, csrf_token=csrf_token, expected_status_int=405)
```
Should we be changing the status code? The '/mailinglistsubscriptionhandler' do accepts the 'put' requests, right?

mon4our replied:
Hey @lkbhitesh07, so the exception raised for this test case is handled by the `handle_exception()` method of core/controllers/base.py, specifically this if block:
```python
if isinstance(exception, TypeError):
            self.error(405)
            values = {
                'error': 'Invalid method %s for %s/ Extra Parameters passed' % (
                    request_method, handler_class_name),
                'status_code': 405
            }
            self._render_exception(values)
            return
``` 
I tried to add another if block inside this one to check if the request method is in the list of methods supported by the handler, which would solve the breaking of this test, but that breaks the code and gives a 500 error for all cases (even the ones where we expect a 405).  Due to this, I replaced the 500 with 405. I also changed the error message of the exception to say 'Invalid method or Extra parameters passed' specifically for this case as I believe the 'name' parameter passed here:
```
{
                    'email': 'email@example.com',
                    'tag': 'Web',
                    'name': 'Name'
                }
``` 
 is extra. This test seems to be an edge case as the changes I have made work for all the other routes and tests.

lkbhitesh07 replied:
Got it, thanks @mon4our

lkbhitesh07 reviewed (changes requested):
Thanks @mon4our for the changes, one inline question.

mon4our commented:
@lkbhitesh07 PTAL

mon4our committed:
Merge branch 'develop' into add-similar-tests-for-no-500-error

lkbhitesh07 reviewed (approved):
Thanks @mon4our LGTM

seanlip commented on core/tests/test_utils.py:
```diff
@@ -3030,12 +3033,35 @@ def get_response_without_checking_for_errors(
+                    url, params=params, expect_errors=True
+                )
+
+            elif http_method == 'POST':
```
This seems a little odd, why would we ever call load_template for POST,PUT, DELETE requests? I thought templates were just the html pages.

mon4our replied:
Hey Sean, I tried putting the `with self.swap()` line inside the `if http_method == 'GET':` block like:

![image](https://github.com/oppia/oppia/assets/120438312/0abbe1bd-3964-438c-bf6b-a52a7e71c748)

So that it is only used for GET requests.
But I got an exception:

![image](https://github.com/oppia/oppia/assets/120438312/f5153116-2075-4e6f-9559-d3cde9102503)

which I tried resolving by defining `response` outside the `if` which also gave an error. It seems the code after `with self.swap()` does not execute if it is kept inside an `if` block. 

So I kept the swap out of the if block.

mon4our replied:
Kindly ignore the previous comment. Successfully implemented your suggestion so that load_template is only called for GET.

mon4our commented:
@seanlip PTAL

mon4our committed:
Made changes so that load_templates is only used for GET

mon4our committed:
Merge branch 'develop' into add-similar-tests-for-no-500-error

mon4our commented:
Hey @seanlip I've addressed your [comment](https://github.com/oppia/oppia/pull/20544#discussion_r1658956201), PTAL. Thanks!

mon4our committed:
Merge branch 'develop' into add-similar-tests-for-no-500-error

seanlip commented on core/controllers/base.py:
```diff
@@ -845,6 +847,16 @@ def handle_exception(
+        if isinstance(exception, TypeError):
+            self.error(405)
+            values = {
+                'error': 'Invalid method %s for %s/ Extra Parameters passed' % (
```
Per the HTTP spec, is "extra parameters passed" a valid condition for surfacing 405 errors? If so, could you please point to some documentation to confirm that?

mon4our replied:
I added "extra parameters passed" for the test in `profile_test.py` below, for more context, you can refer this [review](https://github.com/oppia/oppia/pull/20544#discussion_r1650156797).

mon4our replied:
Done. Removed the 'Extra Parameters passed' part.

seanlip commented on core/controllers/profile_test.py:
```diff
@@ -1168,7 +1168,7 @@ def raise_exception() -> None:
                     'tag': 'Web',
                     'name': 'Name'
-                }, csrf_token=csrf_token, expected_status_int=500)
+                }, csrf_token=csrf_token, expected_status_int=405)
```
@mon4our I don't understand why this is throwing a 405. Isn't PUT in the list of allowed methods for MailingListSubscriptionHandler? What exactly was causing the original 500 error?

mon4our replied:
PUT is indeed allowed for MailingListSubscriptionHandler but the `'name'` parameter passed in this test is extra, which causes a TypeError to be thrown. Since the `handle_exception()` method in `base.py` checks for a TypeError before throwing a 405 status code, the request for this test also gets a 405.
```
test_email_provider_error (core.controllers.profile_test.MailingListSubscriptionHandlerTests) ...
 ERROR:root:Exception raised at http://localhost/mailinglistsubscriptionhandler:
 raise_exception() got an unexpected keyword argument 'name'
 
TypeError: raise_exception() got an unexpected keyword argument 'name'
ERROR:root:Exception raised: raise_exception() got an unexpected keyword argument 'name'
``` 
For more context:
The if condition before raising a 405 status code checks for a TypeError because when I tried sending requests with unsupported methods (like POST/PUT for an endpoint that only supports GET) via postman, the exception raised was always a TypeError.
I was not able to add a check for supported methods inside the if as it was breaking the code.
P.S. I have marked the [earlier review](https://github.com/oppia/oppia/pull/20544#discussion_r1650156797) as unresolved so other reviewers can also see the context for these changes. I'll also add comments in the code to prevent confusion if you agree with these changes.

adhiamboperes replied:
@mon4our, is the "name" parameter required for this request? Wondering if the issue is the endpoint vs the error.

mon4our replied:
@adhiamboperes This seems weird. I removed the 'name' parameter from the `put_json` in the test
![image](https://github.com/user-attachments/assets/5e5c97e4-37f2-40c5-94ee-e0a1ebd4e244)
And on rerunning the test, I still get an exception for 'name'
```
{"error": "raise_exception() got an unexpected keyword argument 'name'", "status_code": 405}
```

mon4our replied:
I just realised, the unexpected keyword argument 'name' is not for the route `/mailinglistsubscriptionhandler` but for the `raise_exception()` method defined in this test itself. I added `*args, **kwargs` to the params of the method so that an exception is not thrown when `raise_exception()` is called (I'm pretty sure it doesn't matter what params are passed to the args, kwargs as we won't be using them anyway)  and replaced the `expected_status_int` to the original 500 so that the test looks like:
![image](https://github.com/user-attachments/assets/b98ef300-5f66-40be-87de-7bfe79b08340)
On running this, we get no errors and all the backend tests pass successfully.

adhiamboperes replied:
It does seem like the swap function is passing the keyword which  `raise_exception` does not accept.
Isn't it sufficient to just have 
```python
def raise_exception(**kwargs) -> None:
    raise Exception('Backend error')
```
as opposed to (*args, **kwargs: True)?

mon4our replied:
I tried:
```python
def raise_exception(**kwargs) -> None:
    raise Exception('Backend error')
``` 
```python
def raise_exception(**kwargs: Any) -> None:
    raise Exception('Backend error')
``` 
as you suggested and also:
```python
def raise_exception(*args) -> None:
    raise Exception('Backend error')
``` 
```python
def raise_exception(*args: Any) -> None:
    raise Exception('Backend error')
``` 
But both approaches didn't work.

When I used :
```python
def raise_exception(*args: Any, **kwargs: Any) -> None:
            raise Exception('Backend error')
``` 
It finally did work, so I pushed this change.
(Mypy doesn't like it when params don't have types, so I used Any)

adhiamboperes replied:
I'll defer to @seanlip for what the correct approach for that should be. Thanks.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -3030,12 +3033,35 @@ def get_response_without_checking_for_errors(
                 params, dict,
                 msg='Expected params to be a dict, received %s' % params)
 
+            if http_method != 'GET':
```
Bit confused about this. I imagine you should just have an assertion at the outset that http_method is one of GET, POST, PUT, DELETE. Why should it be nested in "if params is not None"?

Also, is there some relation that should exist between the http_method and params (e.g. do some methods not support params)? If so we should probably check for that as well.

mon4our replied:
Added assertion that http_method is one of GET, POST, PUT, DELETE and unindented it from under if params is not none. Done.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -3030,12 +3033,35 @@ def get_response_without_checking_for_errors(
+            response = self.testapp.post(
+                url, params=params, expect_errors=True
+            )
+
```
Drop this newline and the one below since all of these are part of the same if/elif/else block.

mon4our replied:
Done.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -3030,12 +3033,35 @@ def get_response_without_checking_for_errors(
+                url, params=params, expect_errors=True
+            )
+
+        elif http_method == 'DELETE':
```
Perhaps change this to an "else" and assert in the next line that http_method is equal to DELETE.

mon4our replied:
Tried that, but since this if block and the one for GET are separate, if the http_method is GET, it is checked again in this if block and this error is thrown:
```
Traceback (most recent call last):
  File "/home/runner/work/oppia/oppia/core/controllers/base_test.py", line 226, in test_that_no_get_results_in_500_error
    self.get_response_without_checking_for_errors(
  File "/home/runner/work/oppia/oppia/core/tests/test_utils.py", line 3060, in get_response_without_checking_for_errors
    self.assertIs(http_method, 'DELETE')
AssertionError: 'GET' is not 'DELETE'
``` 
So, changing it back to elif.

seanlip replied:
OK, then do a "else raise Exception" after this?

mon4our replied:
Done

seanlip reviewed (changes requested):
Thanks @mon4our, sorry for the very late reply on this. I left some comments for you, PTAL. Thanks!

mon4our committed:
Made the requested changes

mon4our committed:
Fixed linting issue

mon4our committed:
Fixed failing backend test, made assertion for http_method neater

mon4our commented:
@seanlip addressed some comments and made the required changes for others. PTAL. Thanks!

mon4our committed:
Merge branch 'develop' into add-similar-tests-for-no-500-error

mon4our committed:
Fixed backend test and made requested changes

mon4our committed:
Fixed Mypy check error

mon4our committed:
Fixed Mypy error by importing Any from typing

mon4our committed:
Fixed linting errors

mon4our committed:
Added comment to address why type Any is used

mon4our committed:
Merge branch 'develop' into add-similar-tests-for-no-500-error

mon4our committed:
Merge branch 'develop' into add-similar-tests-for-no-500-error

mon4our committed:
Merge branch 'develop' into add-similar-tests-for-no-500-error

mon4our commented:
@seanlip friendly ping

seanlip commented on core/controllers/profile_test.py:
```diff
@@ -1153,7 +1153,11 @@ def test_put_function(self) -> None:
 
     def test_email_provider_error(self) -> None:
-        def raise_exception() -> None:
+        # Here we use type Any because we will not be using the
```
I think, for clarity, you'll want to call this something like mock_add_user_to_raise_exception() and define its method signature as "unused_email, unused_tag, name=None". This is because the mock method that you're swapping in should mimic the original method.

I think this should address the 405 issues you're getting.

mon4our replied:
Done

seanlip reviewed (changes requested):
Thanks @mon4our, I'm very sorry for the late review. I left two comments, PTAL.

seanlip commented:
@mon4our Just checking, would you be able to finish this PR?

mon4our commented:
Hey @seanlip I'll make the changes you requested and try to finish the PR by the 30th. Very sorry for the delay.

mon4our committed:
Made requested changes in profile_test

mon4our committed:
Updated comment about mock_add_user_to_raise_exception in profile_test

mon4our committed:
Updated comment about mock_add_user_to_raise_exception in profile_test

mon4our committed:
Made requested changes

mon4our commented:
@seanlip PTAL. Thanks!

mon4our committed:
Merge branch 'develop' into add-similar-tests-for-no-500-error

seanlip reviewed (approved):
@mon4our The changes seem fine, but your basic CI checks are not passing. 

Could you please explain how this happened? Is the pre-push hook not running the lint and mypy checks?

/cc @jnvtnguyen

mon4our committed:
Fixed failing backend, linter and MyPy tests

mon4our committed:
Fixed failing linter tests

mon4our committed:
Fixed failing linter tests

mon4our committed:
Fixed failing MyPy tests

mon4our committed:
Added comment

mon4our committed:
Fixed linting issues

mon4our commented:
@seanlip That is because I used the github editor to make the changes. Very sorry for that. I'll refrain from doing that again. I've fixed all those errors and committed them using standard procedure with the pre-push running successfully. 

Some minor changes:

I've added the type Any [here](https://github.com/oppia/oppia/pull/20544/files#:~:text=unused_email%3A%20Any%2C%20unused_tag%3A%20Any%2C%20name%3A%20Any) as the MyPy type checks were failing.
![image](https://github.com/user-attachments/assets/525d6bb1-7262-49dc-9198-2be70e65706b)


Also added another elif block for GET [here](https://github.com/oppia/oppia/pull/20544/files#:~:text=elif%20http_method%20%3D%3D%20%27GET,pass) as the exception was getting raised everytime the http_method was GET.
![image](https://github.com/user-attachments/assets/cad0cab7-c8fc-4a16-8c9a-d238ffcf1bb8)

mon4our committed:
Merge branch 'develop' into add-similar-tests-for-no-500-error

seanlip commented:
Hi @mon4our, thanks. Some notes:

- Why do you get a mypy error for that? That seems odd and it would be helpful to see more details of what exactly the error is and why it's not solvable using other methods before resorting to using Any.
-  I suggest you drop the added block and just change the last block to `elif not GET` instead of `else`. (Also note that "invalid" is misspelled.)

mon4our commented:
@seanlip 
> * Why do you get a mypy error for that? That seems odd and it would be helpful to see more details of what exactly the error is and why it's not solvable using other methods before resorting to using Any.

I get this error when I try to push the code without using type Any.
```
Starting Mypy type checks.
core/controllers/profile_test.py:1160: error: Function is missing a type annotation for one or more arguments
[no-untyped-def] Found 1 error in 1 file (checked 933 source files)


Mypy type checks unsuccessful. Please fix the errors. For more information,
visit: https://github.com/oppia/oppia/wiki/Backend-Type-Annotations
Push failed, please correct the mypy type annotation issues above.
```

seanlip commented:
@mon4our Then why not use the correct type of the argument instead (e.g. str)? The error seems like it's referring to the absence of any type at all.

mon4our committed:
Added correct types in profile_test

mon4our commented:
> @mon4our Then why not use the correct type of the argument instead (e.g. str)? The error seems like it's referring to the absence of any type at all.

Done

seanlip commented:
Thanks. In the future, please refrain from using `Any` and read the error logs carefully to understand what specific fix needs to be applied.

Did you see my second comment [here](https://github.com/oppia/oppia/pull/20544#issuecomment-2315914974)? I don't think it's been fixed yet.

mon4our committed:
Corrected if block and typo

mon4our commented:
> * I suggest you drop the added block and just change the last block to `elif not GET` instead of `else`. (Also note that "invalid" is misspelled.)

Done

seanlip reviewed (approved):
@mon4our All good. Let's merge this!

seanlip committed:
Merge branch 'develop' into add-similar-tests-for-no-500-error

seanlip commented:
@mon4our You have some CI failures, including coverage -- might be worth taking a look at the logs for those to make sure the checks pass.

mon4our committed:
Merge branch 'develop' into add-similar-tests-for-no-500-error

mon4our committed:
Merge branch 'develop' into add-similar-tests-for-no-500-error

mon4our committed:
Merge branch 'develop' into add-similar-tests-for-no-500-error

mon4our committed:
Added test to fix failing backend coverage

mon4our committed:
Fixed linting issues

mon4our committed:
Fixed Mypy issues

Akhilesh-max committed:
Merge branch 'develop' into add-similar-tests-for-no-500-error
